### PR TITLE
new progress_bar script, now called "animate"

### PR DESCRIPTION
**Pull request by phyver**
Instead of
```````
_mission_init | progress_bar
```````
you now use
```````
export -f _mission_init
animate _mission_init
```````

I hadn't thought about the need to export the function, so it is not very nice. Also, the `animate` script might need to source other things. (It is sourcing `gettext.sh`, but it might need to source `mission_source.sh` or `common.sh`.

The need to use `export -f` makes it definitely non POSIX.

The old `progress_bar` is better in that respect, but it hides the return value of `_mission_init`, so you either need to `set -o pipefail` (non POSIX) or make sure dependency checking is done before calling `_mission_init | progress_bar`.

In the end, I think I prefer the old style version.

### CODE REVIEW

**Comment by phyver** [comment]
What I would like is that the animation runs concurently to `_mission_init` and stops as soon as the `_mission_init` function does.

With the old `progress_bar`, if you don't print enough characters in `_mission_init`, the animation is too slow, and if you print to many of them rapidly, the animation slows the initialization (because we have a `sleep 0.2` before each `read`).
I don't see a nice way of doing that.

**Comment by phyver** [comment]
Slight improvement:
there now is an animate scripts that takes a pid as argument. It displays an animation until the process with given pid stops.
There is a small utility animate.sh you need to source just after the process you want to follow, as in

`````sh
_long_function() {
...
}
_long_function & . animate.sh
`````

I like this solution, but if `_long_function` uses stdout / stderr, the animation will be garbled...

**Comment by phyver** [comment]
What's really missing from POSIX is the ability to pass arguments when sourcing a file.
This is supported by bash and zsh. I did a third version where you can use

`````sh
_long_function() {
...
}
. animate.sh _long_function
`````

It redirects stdout to `/dev/null` and stderr to a temporary file that is displayed at the end.
The return value is the correct one.

To use with a pure POSIX shell, you need to set the arguments `$@` by hand with

`````sh
_long_function() {
...
}
set _mission_init
. animate.sh _mission_init
`````

That might well be the version I prefer.

**Comment by phyver** [comment]
@rlepigre
What do you think of the current version of `progress_bar` / `progress_bar.sh`?
Do you prefer the original version (pro: simple filter, cons: problems with stderr and return values) or the new version (pro: correct return value, and stderr displayed at the end, cons: must be sourced to get correct environment)?

File: bin/progress_bar.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    design=$2
+    shift 2
+  fi
+  LOG=$(mktemp)
+  "$@" 1>/dev/null 2>"$LOG" &
+  _PID=$!
+  progress_bar -b "$design" $_PID
+  cat "$LOG"
+  rm -f "$LOG"
+  unset LOG
```
Comment (rlepigre) [inline bin/progress_bar.sh:37]: Should other variables like `design` also be unset?

Maybe we should avoid using the name `LOG` here, it sounds a bit generic and could potentially clash with an environment variable, don't you think?

File: bin/progress_bar [hunk 2]
```diff
@@ -75,119 +131,99 @@ progress_bar () {
...
   do
-    echo -en "${STR[$COUNT]}"
-    COUNT=$(((COUNT+1)%L))
-    # Slow down the animation a little bit.
-    sleep .1
+    for s in "${STR[@]}"
+    do
+      printf "$s"
+      # Slow down the animation a little bit.
+      sleep .2
```
Comment (rlepigre) [inline bin/progress_bar:185]: In the current version the animation is really slow. I tried `sleep .1` and that seems better for most designs.
Actually, maybe we should make the "sleep time" a parameter of the design?

**Review by rlepigre — COMMENTED** [review]
> What do you think of the current version of progress_bar / progress_bar.sh?
> Do you prefer the original version (pro: simple filter, cons: problems with stderr and return values) or the new version (pro: correct
> return value, and stderr displayed at the end, cons: must be sourced to get correct environment)?

The current version seems great, I only have a few comments.

Also, there is one thing that is a bit annoying: the shell ends up printing a kind of status line, I'm not sure where it comes from.
Here is an example of what I am talking about (line 3):
```
[mission 8] $ gsh reset
While you are waiting, a caterpillar crawls by...
[1]+  Done                    "$@" > /dev/null 2> "$LOG"

  |                                |  
--+--------------------------------+--
  | Use the command                |  
  |     $ gsh goal                 |  
  | to discover your next mission. |  
--+--------------------------------+--
  |                                |  

[mission 8] $ 
```

File: bin/progress_bar (hunk 2 above)
Comment (phyver) [reply bin/progress_bar:185]: I almost made it depend on the design, to have a slow snail!

File: bin/progress_bar.sh (hunk 1 above)
Comment (phyver) [reply bin/progress_bar.sh:37]: I don't think there's a way to unset `_PID` though...

**Comment by phyver** [comment]
Is that better?
Can you test a few things to check everything is OK?

File: bin/progress_bar.sh (hunk 1 above)
Comment (rlepigre) [reply bin/progress_bar.sh]: OK, that's fine I guess since it start with an underscore.

**Comment by rlepigre** [comment]
Everything looks good, I think we can merge this!
It might be worth squashing the branch, which you can do in the web interface be choosing the "squash and merge" option on the merge button. If you do that it will let you change the commit message.

**Comment by phyver** [comment]
> It might be worth squashing the branch, which you can do in the web interface be choosing the "squash and merge" option on the merge button. If you do that it will let you change the commit message.
>
OK, that didn't go as planned. Next time, I'll do it the old-fashioned way.

The title of the commit was wrong, and it messed up a local branch.
I forced push an update to correct that...

I'm still not completely at ease with merging / rebasing multiple local branches and keeping a somewhat clean history...

**Pull request merged**